### PR TITLE
Create goroutines adaptively during metrics gathering

### LINKED ISSUE #369: Find tradeoff for number of goroutines during `Gather`
https://github.com/prometheus/client_golang/blob/master/prometheus/registry.go#L381 starts one goroutine per registered `Collector`. Cases of 1M registered collectors have been reported. With fast collections, the overhead of starting a goroutine isn't really worth it, and there is a notable memory peak caused by it. Let's come up with a way to adaptively find a tradeoff. With fast collectors, one goroutine is fine. Even with slow ones, starting thousands of goroutine will have diminishing returns.

### PR DESCRIPTION
**Pull request by beorn7**
Attempt to fix #369. This needs to be tested with one of those scenarios as in https://github.com/prometheus/statsd_exporter/issues/76 .

@grobie @brian-brazil

### CODE REVIEW

**Comment by matthiasr** [comment]
prometheus/statsd_exporter#119 has benchmarks and profiling that you can use to test the effects of this.

**Comment by brian-brazil** [comment]
The code seems okay from a first lookthrough. I'd worry that this approach might cause unpredictable goroutine/memory usage when a varying number of collectors are slow. Maybe limit the number of goroutines to the number of cores or something to bound it?

**Comment by beorn7** [comment]
@matthiasr I'll check out the statsd_exporter benchmarks as a test case for this.

@brian-brazil the number of goroutine is effectively limited here to the number of collectors. Limiting it to the number of cores would not make sense. It's the whole point of the go execution model to have more goroutines than cores. E.g. if a goroutine is waiting for I/O, it gets effectively detached from its current OS thread, and the latter will run another goroutine.

**Comment by beorn7** [comment]
Benchmark result from https://github.com/prometheus/statsd_exporter/pull/119 :

BEFORE:
```
BenchmarkGatherGauge/m_100000_l_1-4                    2         634044750 ns/op        61646164 B/op    1157861 allocs/op
BenchmarkGatherGauge/m_10000_l_10-4                    3         342434786 ns/op        26507605 B/op     514057 allocs/op
BenchmarkGatherGauge/m_10_l_10000-4                    5         296876316 ns/op        23959355 B/op     404184 allocs/op
BenchmarkGatherGauge/m_1_l_100000-4                    3         496134591 ns/op        24761392 B/op     404040 allocs/op
BenchmarkGatherCounter/m_100000_l_1-4                  2         550082110 ns/op        61584880 B/op    1157420 allocs/op
BenchmarkGatherCounter/m_10000_l_10-4                  3         354095715 ns/op        26505234 B/op     514031 allocs/op
BenchmarkGatherCounter/m_10_l_10000-4                  5         294096041 ns/op        23966792 B/op     404187 allocs/op
BenchmarkGatherCounter/m_1_l_100000-4                  3         481249290 ns/op        24753322 B/op     404016 allocs/op
BenchmarkGatherSummary/m_100000_l_1-4                  1        1502250708 ns/op        105471992 B/op   2606115 allocs/op
BenchmarkGatherSummary/m_10000_l_10-4                  1        1237115864 ns/op        60737560 B/op    1833445 allocs/op
BenchmarkGatherSummary/m_10_l_10000-4                  2         603992947 ns/op        53568084 B/op    1604204 allocs/op
BenchmarkGatherSummary/m_1_l_100000-4                  1        1067372223 ns/op        57537016 B/op    1703965 allocs/op
BenchmarkGatherHistogram/m_100000_l_1-4                1        1748068925 ns/op        156688928 B/op   4805885 allocs/op
BenchmarkGatherHistogram/m_10000_l_10-4                2         585414979 ns/op        110880208 B/op   4018966 allocs/op
BenchmarkGatherHistogram/m_10_l_10000-4                2         538948882 ns/op        107956244 B/op   3904172 allocs/op
BenchmarkGatherHistogram/m_1_l_100000-4                2         704194725 ns/op        108760040 B/op   3904024 allocs/op
```

AFTER (i.e. with this PR in):
```
BenchmarkGatherGauge/m_100000_l_1-4                    3         440166007 ns/op        52157970 B/op    1009594 allocs/op
BenchmarkGatherGauge/m_10000_l_10-4                   10         172004623 ns/op        25938732 B/op     504239 allocs/op
BenchmarkGatherGauge/m_10_l_10000-4                    5         263151519 ns/op        23970416 B/op     404233 allocs/op
BenchmarkGatherGauge/m_1_l_100000-4                    3         482181690 ns/op        24759512 B/op     404018 allocs/op
BenchmarkGatherCounter/m_100000_l_1-4                  3         437489358 ns/op        52168122 B/op    1009618 allocs/op
BenchmarkGatherCounter/m_10000_l_10-4                 10         182594454 ns/op        25941801 B/op     504233 allocs/op
BenchmarkGatherCounter/m_10_l_10000-4                  3         349815703 ns/op        23961765 B/op     404202 allocs/op
BenchmarkGatherCounter/m_1_l_100000-4                  3         483724620 ns/op        24746320 B/op     404000 allocs/op
BenchmarkGatherSummary/m_100000_l_1-4                  1        1298970300 ns/op        84972176 B/op    2309705 allocs/op
BenchmarkGatherSummary/m_10000_l_10-4                  5         286344748 ns/op        55537083 B/op    1704213 allocs/op
BenchmarkGatherSummary/m_10_l_10000-4                  3         446114870 ns/op        53563317 B/op    1604220 allocs/op
BenchmarkGatherSummary/m_1_l_100000-4                  2         883968198 ns/op        54331852 B/op    1603902 allocs/op
BenchmarkGatherHistogram/m_100000_l_1-4                1        1009265108 ns/op        136181704 B/op   4509751 allocs/op
BenchmarkGatherHistogram/m_10000_l_10-4                3         466298888 ns/op        109941794 B/op   4004212 allocs/op
BenchmarkGatherHistogram/m_10_l_10000-4                2         558274380 ns/op        107960956 B/op   3904228 allocs/op
BenchmarkGatherHistogram/m_1_l_100000-4                2         891237764 ns/op        108757552 B/op   3903994 allocs/op
```

The cases with many individual collectors to gather from (`m_100000`) are treated more efficiently but not dramatically so. @matthiasr told me it gets interesting with a million and more, but that doesn't work with summaries (OOMs). Will run that for a Counter only. Also, looking at the memory profile is also interesting.

**Comment by beorn7** [comment]
Counter with a million:

BEFORE:
```
BenchmarkGatherCounter/m_1000000_l_1-4                 1        11464764011 ns/op       1236364384 B/op 14072161 allocs/op
BenchmarkGatherCounter/m_1_l_1000000-4                 1        7028819377 ns/op        263177648 B/op   4038419 allocs/op
```

AFTER:
```
BenchmarkGatherCounter/m_1000000_l_1-4                 1        4894638356 ns/op        599587720 B/op  10076272 allocs/op
BenchmarkGatherCounter/m_1_l_1000000-4                 1        7299555268 ns/op        263168624 B/op   4038438 allocs/op
```

Interestingly, gathering from 1M collectors is now even faster than gathering from one collector with a million metrics.

**Comment by beorn7** [comment]
Profiles to look at:

[m_1_l_1000000.cpu.BEFORE.pb.gz](https://github.com/prometheus/statsd_exporter/files/1674498/m_1_l_1000000.cpu.BEFORE.pb.gz)
[m_1_l_1000000.cpu.pb.gz](https://github.com/prometheus/statsd_exporter/files/1674499/m_1_l_1000000.cpu.pb.gz)
[m_1_l_1000000.mem.BEFORE.pb.gz](https://github.com/prometheus/statsd_exporter/files/1674500/m_1_l_1000000.mem.BEFORE.pb.gz)
[m_1_l_1000000.mem.pb.gz](https://github.com/prometheus/statsd_exporter/files/1674501/m_1_l_1000000.mem.pb.gz)
[m_1000000_l_1.cpu.BEFORE.pb.gz](https://github.com/prometheus/statsd_exporter/files/1674502/m_1000000_l_1.cpu.BEFORE.pb.gz)
[m_1000000_l_1.cpu.pb.gz](https://github.com/prometheus/statsd_exporter/files/1674503/m_1000000_l_1.cpu.pb.gz)
[m_1000000_l_1.mem.BEFORE.pb.gz](https://github.com/prometheus/statsd_exporter/files/1674504/m_1000000_l_1.mem.BEFORE.pb.gz)
[m_1000000_l_1.mem.pb.gz](https://github.com/prometheus/statsd_exporter/files/1674505/m_1000000_l_1.mem.pb.gz)

Bottom line: With the changes here, the profile looks broadly the same, whether you collect 1M collectors with one metric or one collector with 1M metrics.

**Comment by beorn7** [comment]
Just checked how many goroutines were started for gathering with this change:
To collect from one collector, indeed only one.
To collect from 1M collectors, it started 11 goroutines.

This seems to work out quite nicely.

**Comment by beorn7** [comment]
I added a safety measure to not schedule a large or even infinite number of goroutines (which can happen in theory).

With the benchmark results in, I think this works great and we should merge. @matthiasr WDYT?

File: prometheus/registry.go [hunk 1]
```diff
@@ -368,22 +376,12 @@ func (r *Registry) Gather() ([]*dto.MetricFamily, error) {
 	)
 
 	r.mtx.RLock()
+	goroutineBudget := len(r.collectorsByID)
```
Comment (matthiasr) [inline prometheus/registry.go]: is there a reasonable maximum that we could put here to limit memory usage? Something like "never start more than 100k goroutines" (~ 40MB) so that in the pathological case of many slow collectors, the collection time suffers but we don't overload the system?

Comment (beorn7) [reply prometheus/registry.go]: I don't think there is a one-size-fits-it-all maximum. If you have a large system, 1M goroutines might be what you want. If you are on a small system, you want fewer, but if you have 1M slow collectors on a small system, you are doomed anyway.

Comment (beorn7) [reply prometheus/registry.go]: Or in other words: I find it reasonable to start one goroutine per slow collector. And that's what this PR accomplishes in an adaptive fashion.
